src/Disks/IO/ReadBufferFromAzureBlobStorage.cpp

```diff
@@ -87,46 +87,16 @@ bool ReadBufferFromAzureBlobStorage::nextImpl()
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Attempt to read beyond right offset ({} > {})", offset, read_until_position - 1);
     }
 
-    if (!initialized)
-        initialize();
-
     if (use_external_buffer)
     {
         data_ptr = internal_buffer.begin();
         data_capacity = internal_buffer.size();
     }
 
-    size_t to_read_bytes = std::min(static_cast<size_t>(total_size - offset), data_capacity);
-    size_t bytes_read = 0;
-
-    size_t sleep_time_with_backoff_milliseconds = 100;
+    if (static_cast<size_t>(offset) >= getFileSize())
```

Comment:
If it is a azure disk read it will be better not to execute additional request to azure just to get file size (getFileSize is only used for external table engines) as it is already known before we create the read buffer. So may be just pass file_size as optional argument in constructor of `ReadBufferFrpmAzureBlob`? 

btw why do we need this check? There is no such check for other buffers AFAIK

Reply:
> In getFileSize() function we only read the value once, it is just returned for future calls.

Yes, I saw that it is fetched only once, but if it is avoidable - better to avoid, we already have a file size when we create this read buffer, better to use it.

> For other buffers, I see we send the request to read the full intended length of the file.

But isn't it the same as here? Here we have a buffer and try to read `data_capacity` size to it, but if we reach the end of file while reading into buffer - we will just read nothing and the buffer will be empty, so we return `false` as well. Or am I missing something?

Reply:
> But isn't it the same as here? Here we have a buffer and try to read data_capacity size to it, but if we reach the end of file while reading into buffer - we will just read nothing and the buffer will be empty, so we return false as well. Or am I missing something?

DownloadTo function takes Range as a parameter (offset, length) if length is zero it reads full file, so when we reach end of file, the range will be (offset=file_size, length), even if length is zero we see error message that the "The range specified is invalid for the current size of the resource". 
Previously when we used to Download & then ReadToBuffer, there we had a check to_read_bytes = min(total_size-offset, length), and we would just read 0 bytes & return false.
Hope that clarifies the need. Please suggest if you have better options.

Reply:
But `read_until_position` is not always set. It is not set for `remote_filesystem_read_method=read`